pageserver/src/tenant/timeline.rs

```diff
@@ -4875,11 +4875,15 @@ impl<'a> TimelineWriter<'a> {
 
         // Rolling the open layer can be triggered by:
         // 1. The distance from the last LSN we rolled at. This bounds the amount of WAL that
-        //    the safekeepers need to store.
+        //    the safekeepers need to store.  For sharded tenants, we multiply by shard count to
+        //    account for how writes are distributed across shards: we expect each node to consume
+        //    1/count of the LSN on average.
         // 2. The size of the currently open layer.
         // 3. The time since the last roll. It helps safekeepers to regard pageserver as caught
         //    up and suspend activity.
-        if distance >= self.get_checkpoint_distance().into() {
+        if distance
+            >= self.get_checkpoint_distance() as i128 * self.shard_identity.count.count() as i128
```

Comment:
The obvious alternative to this heuristic would be to simply get rid of the LSN-based condition, and rely on time condition and layer size condition only.

That would probably work in practice, but might store up future trouble: it would result in very high safekeeper disk requirement in the edge case of a timeline writing at a high rate to pages that are only on a subset of shards, and it would also be a risk if we choose to adjust checkpoint timeout in future (this would amplify the potential disk usage on safekeeper)